### PR TITLE
feat: update CompactWads serialization with colon-separated format

### PR DESCRIPTION
**Pull request by big14way**
# Multi-Mint Token Concatenation Implementation

## Issue Resolution

This PR resolves the implementation of multi-mint token concatenation to align with the Cashu specification for colon-separated token format, as discussed in [cashubtc/nuts#275](https://github.com/cashubtc/nuts/issues/275).

## Background

Previously, our implementation used a CBOR array approach for handling multiple tokens from different mints. However, the Cashu community has moved towards standardizing on a colon-separated format for better interoperability and simpler parsing across different implementations.

### 🔄 **Token Serialization Format**

**Previous Approach (CBOR Array):**
- Multiple tokens stored as CBOR-encoded array
- Complex serialization/deserialization logic
- Less human-readable format

**New Approach (Colon-Separated):**
- Multiple tokens joined with `:` separator
- Simple string concatenation: `cashuBtoken1:cashuBtoken2:cashuBtoken3`
- Human-readable and easier to debug

### 🏗️ **Parsing Logic Simplification**

**Core Implementation:**
```rust
// Clean, specification-compliant parsing
let token_strings: Vec<&str> = s.split(':').collect();
let mut wads = Vec::with_capacity(token_strings.len());
for token_str in token_strings {
    let wad = CompactWad::from_str(token_str)?;
    wads.push(wad);
}
```

**Benefits:**
- Follows Cashu specification exactly
- Simpler logic with fewer edge cases
- Better error handling and validation
- Consistent behavior across implementations

#### Token Configuration Tests
- **Single Token**: Validates basic functionality without separators
- **Multiple Tokens**: Tests various combinations of single/multi-proof tokens
- **Scalability**: Ensures proper handling of 2, 3, and more tokens

#### Error Handling Tests
- **Malformed Separators**: Double colons, leading/trailing colons
- **Invalid Tokens**: Mixed valid/invalid tokens, missing prefixes
- **Edge Cases**: Concatenated tokens without separators

#### Roundtrip Validation
All tests verify complete serialization → deserialization → content verification cycles to ensure data integrity.

### 🎯 **Format Specification Compliance**

**Single Token:**
```
cashuBbase64encodeddata
```

**Multiple Tokens:**
```
cashuBtoken1:cashuBtoken2:cashuBtoken3
```

**Validation Rules:**
- Each token must start with `cashuB` prefix
- Tokens separated by exactly one `:` character
- No leading or trailing colons allowed
- No empty tokens between separators

### 🔧 **Implementation Details**

**Serialization:**
- Converts each `CompactWad` to string individually
- Joins multiple tokens with `:` separator
- Maintains existing single-token format for compatibility

**Deserialization:**
- Splits input string on `:` character
- Parses each segment as individual token
- Validates all tokens before accepting the collection
- Fails fast on any invalid token

### ✅ **Forward Compatibility**
- New format works with existing single-token implementations
- Gradual adoption possible across the ecosystem
- No breaking changes for single-token use cases

### 🚫 **Backward Compatibility Removal**
- Removed CBOR array fallback parsing
- Clean implementation focused on specification compliance
- Simplified codebase without legacy support burden

### 🎯 **Specification Alignment**
- Full compliance with emerging Cashu standards
- Consistent behavior across different wallet implementations
- Better interoperability in the Cashu ecosystem

### 🛠️ **Developer Experience**
- Simpler debugging with human-readable format
- Easier testing and validation
- Reduced complexity in parsing logic

### 🔒 **Reliability**
- Comprehensive error handling for malformed inputs
- Clear validation rules and error messages
- Robust roundtrip testing ensures data integrity

### 📈 **Performance**
- Faster parsing with simple string operations
- Reduced memory overhead compared to CBOR
- More efficient serialization for multiple tokens

### ✅ **Positive Test Cases**
1. **Single Proof Token**: Basic functionality validation
2. **Multiple Proofs Token**: Complex token structure handling
3. **Mixed Token Types**: Different proof counts per token
4. **Scalability**: Multiple tokens (2, 3, N tokens)

### ❌ **Error Cases**
1. **Format Violations**: Invalid separators, malformed structure
2. **Token Validation**: Invalid prefixes, corrupted data
3. **Edge Cases**: Empty inputs, boundary conditions

### 🔄 **Integration Testing**
- Full serialization/deserialization cycles
- Cross-compatibility with existing token formats
- Performance validation for large token collections

## Future Considerations

This implementation provides a solid foundation for:
- Additional token metadata support
- Enhanced multi-mint workflows
- Improved user experience in multi-mint scenarios
- Better integration with Cashu ecosystem tools

The colon-separated format aligns with community standards and ensures our implementation remains compatible with the broader Cashu ecosystem as specifications evolve.

### CODE REVIEW

**Review by tdelabro — CHANGES_REQUESTED** [review]
Overall, not a big fan of your tests. It's not clear what you want to test. And you often test twice the same thing.

Tests we want:
ok test:
- create a token with 1 proof, compact it, to string, from string, uncompact, assert it is the same content
- same thing with a token with multiple proofs
- same thing but with two tokens, 1 of 1 proof, 1 with multiple proofs
- same thing with 3 tokens

ko tests:
- wad string of two tokens not separated by `:`
- wad string with `::`
- wad string starting with ":"
- wad string finishing by ":"
- wad with a valid and an invalid token
- wad string where one token has no `cashuB` prefix

**Comment by tdelabro** [comment]
> # PKI Attestation Implementation
> ## Overview
> 
> The PKI Attestation system is a comprehensive smart contract implementation for validating vTPM attestation tokens using Public Key Infrastructure (PKI). This implementation provides on-chain verification of attestation tokens with a complete certificate chain validation process.
> ## Key Features
> 
>     * JWT token validation with RS256 algorithm support
> 
>     * X.509 certificate chain validation (leaf → intermediate → root)
> 
>     * DER decoding and certificate parsing
> 
>     * Certificate validity period checks
> 
>     * RSA signature verification using OpenZeppelin's RSA library
> 
>     * Root certificate fingerprint validation
> 
>     * JWT payload validation (audience, issuer, expiration)
> 
> 
> ## Contract Components
> ### Core Structures
> 
>     1. **Certificate**
>        
>        * DER-encoded certificate bytes
>        * TBS (To Be Signed) certificate data
>        * Public key components (modulus and exponent)
>        * Validity period
>        * Fingerprint
>        * Signature
> 
>     2. **JWT Components**
>        
>        * Header (algorithm, x5c chain)
>        * Payload (audience, issuer, expiration, etc.)
> 
> 
> ### Validation Process
> 
> The attestation verification process follows these steps:
> 
>     1. **Certificate Chain Length Validation**
>        
>        * Ensures exactly 3 certificates are provided (leaf, intermediate, root)
> 
>     2. **JWT Header Validation**
>        
>        * Verifies RS256 algorithm is used
>        * Extracts x5c certificate chain
> 
>     3. **Certificate Parsing**
>        
>        * Decodes base64 certificates
>        * Extracts TBS certificate data
>        * Parses RSA public keys
>        * Extracts validity periods
>        * Calculates fingerprints
> 
>     4. **Certificate Validity Checks**
>        
>        * Validates certificate dates
>        * Ensures minimum RSA key size (2048 bits)
>        * Verifies certificate chain signatures
>        * Validates root certificate fingerprint
> 
>     5. **JWT Validation**
>        
>        * Verifies JWT signature using leaf certificate
>        * Validates payload claims (audience, issuer)
>        * Checks token expiration
> 
> 
> ### Error Handling
> 
> Comprehensive error reporting through:
> 
>     * Detailed error messages
> 
>     * Granular error types
> 
>     * Event emission for tracking
> 
>     * Validation failure events with specific error codes
> 
> 
> ### Security Features
> 
>     1. **Upgradability**
>        
>        * UUPS upgradeable pattern
>        * Owner-controlled upgrades
>        * Safe initialization pattern
> 
>     2. **Access Control**
>        
>        * Owner-managed root certificate updates
>        * Protected upgrade functionality
> 
>     3. **Cryptographic Security**
>        
>        * Minimum 2048-bit RSA key requirement
>        * SHA-256 for fingerprints
>        * Strict certificate chain validation
> 
> 
> ## Technical Details
> ### Constants
> 
> ```solidity
> string public constant REQUIRED_ALGORITHM = "RS256";
> string public constant REQUIRED_AUDIENCE = "https://sts.google.com";
> string public constant REQUIRED_ISSUER = "https://confidentialcomputing.googleapis.com";
> uint256 public constant REQUIRED_CERT_COUNT = 3;
> uint256 public constant MIN_RSA_KEY_SIZE = 256; // 2048 bits = 256 bytes
> ```
> 
> ### Events
> 
>     * `ValidationSuccess`: Emitted on successful attestation verification
> 
>     * `ValidationFailure`: Emitted on validation failures with error details
> 
>     * `RootFingerprintUpdated`: Tracks root certificate fingerprint changes
> 
> 
> ### Error Types
> 
> Comprehensive error enumeration covering:
> 
>     * JWT format and validation errors
> 
>     * Certificate parsing and validation errors
> 
>     * Cryptographic verification failures
> 
>     * Timing and expiration issues
> 
> 
> ## Usage
> 
>     1. **Contract Deployment**
>        ```solidity
>        function initialize(address initialOwner, bytes32 _trustedRootFingerprint)
>        ```
> 
>     2. **Attestation Verification**
>        ```solidity
>        function verifyAttestation(
>            bytes memory headerB64,
>            bytes memory payloadB64,
>            bytes[] memory x5cChain
>        ) external returns (bytes memory payload)
>        ```
> 
>     3. **Root Certificate Management**
>        ```solidity
>        function updateTrustedRootFingerprint(bytes32 newFingerprint) external onlyOwner
>        ```
> 
> 
> ## Implementation Notes
> 
>     * The implementation includes simplified versions of complex operations (ASN.1 parsing, base64 decoding) that should be replaced with production-ready implementations
> 
>     * Certificate parsing is currently simplified and needs to be enhanced for production use
> 
>     * JSON parsing is minimal and should be expanded for production deployment
> 
>     * RSA validation uses OpenZeppelin's cryptography library for secure operations
> 
> 
> ## Future Improvements
> 
>     1. Implement full ASN.1 parsing for certificate handling
> 
>     2. Add comprehensive base64 decoding
> 
>     3. Enhance JSON parsing capabilities
> 
>     4. Add more detailed validation reporting
> 
>     5. Implement caching mechanisms for performance optimization
> 
>     6. Add support for certificate revocation checking
> 
>     7. Enhance error reporting with more detailed information
>        closes [fix multi-mint token concatenation #156](https://github.com/nutty-raccoon/paynet/issues/156)

This is not related to this issue.

**Comment by big14way** [comment]
@tdelabro i just reviewed it again thanks for the feedback you can check again and let me know

File: crates/libs/wallet/src/types/compact_wad.rs [hunk 1]
```diff
@@ -124,27 +126,28 @@ impl<U: Unit + DeserializeOwned> FromStr for CompactWad<U> {
...
-            .ok_or(Error::UnsupportedWadFormat)?;
+        let token_strings: Vec<&str> = s.split(':').collect();
 
-        let decode_config = general_purpose::GeneralPurposeConfig::new()
-            .with_decode_padding_mode(bitcoin::base64::engine::DecodePaddingMode::Indifferent);
-        let decoded = GeneralPurpose::new(&alphabet::URL_SAFE, decode_config).decode(s)?;
-        let token = ciborium::from_reader(&decoded[..])?;
-        Ok(token)
+        let mut wads = Vec::with_capacity(token_strings.len());
+        for token_str in token_strings {
```
Comment (tdelabro) [inline crates/libs/wallet/src/types/compact_wad.rs:144]: you are collecting twice here when you actually just need to do it once:

```rust

let wads = s.split(':')
   .map(|tokent_str| CompactWad::from_str(token_str))
   .collect::<Result<Vec<_>, _>>()
   .map_err(|e| Error::InvalidWadToken(Box::new(e)))?;;

        Ok(CompactWads(wads))
```

should work

File: crates/libs/wallet/src/types/compact_wad.rs [hunk 2]
```diff
@@ -225,3 +228,282 @@ where
...
+    use nuts::nut01::PublicKey;
+    use nuts::nut02::KeysetId;
+    use nuts::{Amount, traits::Unit};
+    use std::str::FromStr;
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+    #[serde(rename_all = "lowercase")]
+    pub enum TestUnit {
+        Sat,
+    }
```
Comment (tdelabro) [inline crates/libs/wallet/src/types/compact_wad.rs:245]: there are already TestUnit available for you to use in crates/libs/nuts/src/lib.rs

Use it instead of recreating a new one with the whole implementation

Comment (tdelabro) [reply crates/libs/wallet/src/types/compact_wad.rs:245]: I told you in https://github.com/nutty-raccoon/paynet/pull/169/files#r2232352222

**Comment by big14way** [comment]
@tdelabro yes i just did that  did that you can re-confirm again

**Review by tdelabro — DISMISSED** [review]
